store/blockstore.go

```diff
@@ -0,0 +1,122 @@
+package store
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"sync"
+
+	"github.com/lazyledger/optimint/types"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/sha3"
+)
+
+var (
+	blockPreffix = [1]byte{1}
+	indexPreffix = [1]byte{2}
+)
+
+type DefaultBlockStore struct {
+	db KVStore
+
+	height uint64
+
+	mtx sync.RWMutex
+}
+
+var _ BlockStore = &DefaultBlockStore{}
+
+func NewBlockStore() BlockStore {
+	return &DefaultBlockStore{db: NewInMemoryKVStore()}
+}
+
+func (bs *DefaultBlockStore) Height() uint64 {
+	bs.mtx.RLock()
+	defer bs.mtx.RUnlock()
+	return bs.height
+}
+
+func (bs *DefaultBlockStore) SaveBlock(block *types.Block) error {
+	// TODO(tzdybal): proper serialization & hashing
+	hash, err := getHash(block)
+	if err != nil {
+		return err
+	}
+	key := append(blockPreffix[:], hash[:]...)
+
+	height := make([]byte, 8)
+	binary.LittleEndian.PutUint64(height, block.Header.Height)
+	ikey := append(indexPreffix[:], height[:]...)
+
+	var value bytes.Buffer
+	enc := gob.NewEncoder(&value)
+	err = enc.Encode(block)
+	if err != nil {
+		return err
+	}
+
+	bs.mtx.Lock()
+	defer bs.mtx.Unlock()
+	// TODO(tzdybal): use transaction for consistency of DB
+	err = multierr.Append(err, bs.db.Set(key, value.Bytes()))
+	err = multierr.Append(err, bs.db.Set(ikey, hash[:]))
+
+	if block.Header.Height > bs.height {
+		bs.height = block.Header.Height
+	}
+
+	return err
+}
+
+// TODO(tzdybal): what is more common access pattern? by height or by hash?
+// currently, we're indexing height->hash, and store blocks by hash, but we might as well store by height
+// and index hash->height
```

Comment:
I think we need both. In optimint you can dictate the access pattern, on the app-side (e.g. via RPC), you can't really know (but you could look into some typical cosmos "apps").